apollo-router/src/plugins/authentication.rs

```diff
@@ -189,6 +204,136 @@ async fn get_jwks(url: Url) -> Option<JwkSet> {
+                if purpose != &PublicKeyUse::Signature {
+                    continue;
+                }
+            } else if let Some(purpose) = &key.common.key_operations {
+                if !purpose.contains(&KeyOperations::Verify) {
+                    continue;
+                }
+            } else {
+                continue;
+            };
+
+            // Let's see if we have a specified kid and if they match
+            if criteria.kid.is_some() && key.common.key_id == criteria.kid {
+                key_score += 1;
+            }
+
+            // Furthermore, we would like our algorithms to match, or at least the kty
+            // If we have an algorithm that matches, boost the score
+            match key.common.algorithm {
+                Some(algorithm) => {
+                    if algorithm != criteria.alg {
+                        continue;
+                    }
+                    key_score += 1;
+                }
+                // If a key doesn't have an algorithm, then we match the "alg" specified in the
+                // search criteria against all of the algorithms that we support.  If the
+                // key.algorithm parameters match the type of parameters for the "family" of the
+                // criteria "alg", then we update the key to use the value of "alg" provided in
+                // the search criteria.
+                // If not, then this is not a usable key for this JWT
+                // Note: Matching algorithm parameters may seem unusual, but the appropriate
+                // algorithm details are not structured for easy consumption in jsonwebtoken and
+                // this is the simplest way to determine algorithm family.
+                None => match criteria.alg {
+                    Algorithm::HS256 | Algorithm::HS384 | Algorithm::HS512 => {
+                        if let AlgorithmParameters::OctetKey(_) = key.algorithm {
+                            key.common.algorithm = Some(criteria.alg);
+                        } else {
+                            continue;
+                        }
+                    }
+                    Algorithm::RS256
+                    | Algorithm::RS384
+                    | Algorithm::RS512
+                    | Algorithm::PS256
+                    | Algorithm::PS384
+                    | Algorithm::PS512 => {
+                        if let AlgorithmParameters::RSA(_) = key.algorithm {
+                            key.common.algorithm = Some(criteria.alg);
+                        } else {
+                            continue;
+                        }
+                    }
+                    Algorithm::ES256 | Algorithm::ES384 | Algorithm::EdDSA => {
+                        if let AlgorithmParameters::EllipticCurve(_) = key.algorithm {
+                            key.common.algorithm = Some(criteria.alg);
+                        } else {
+                            continue;
+                        }
```

Comment:
```suggestion
                    Algorithm::ES256 => {
                      if let AlgorithmParameters::EllipticCurve(params) = key.algorithm {
                        if params.curve == EllipticCurve::P256 {
                          key.common.algorithm = Some(criteria.alg);
                        } else {
                          continue;
                        }
                      }
                      continue;
                    },
                     Algorithm::ES384 => {
                      if let AlgorithmParameters::EllipticCurve(params) = key.algorithm {
                        if params.curve == EllipticCurve::P384 {
                          key.common.algorithm = Some(criteria.alg);
                        } else {
                          continue;
                        }
                      }
                      continue;
                    },
                     Algorithm::EdDSA => {
                      if let AlgorithmParameters::EllipticCurve(params) = key.algorithm {
                        if params.curve == EllipticCurve::Ed25519 {
                          key.common.algorithm = Some(criteria.alg);
                        } else {
                          continue;
                        }
                      }
                      continue;
                    },
```

The `256` and `384` refer to curves, not hash size

Reply:
done: 53b6295c1934a75d51e24fcb8f198576af8f7fb4